arangod/RestServer/SoftShutdownFeature.cpp

```diff
@@ -0,0 +1,234 @@
+        << "Received second soft shutdown request, ignoring it...";
+    return;
+  }
+
+  LOG_TOPIC("fedd2", INFO, Logger::STARTUP)
+      << "Initiating soft shutdown...";
+
+  // Tell asynchronous job manager:
+  auto& generalServerFeature = _server.getFeature<GeneralServerFeature>();
+  auto& jobManager = generalServerFeature.jobManager();
+  jobManager.initiateSoftShutdown();
+
+  // Tell Pregel subsystem:
+  auto& pregelFeature = _server.getFeature<pregel::PregelFeature>();
+  pregelFeature.initiateSoftShutdown();
+
+  // And initiate our checker to watch numbers:
+  if (!::queueShutdownChecker(_workItemMutex, _workItem, _checkFunc)) {
+    // Make it hard in this case:
+    LOG_TOPIC("de425", INFO, Logger::STARTUP)
+        << "Failed to queue soft shutdown checker, doing hard shutdown "
+           "instead.";
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+    _server.beginShutdown();
+  }
+}
+    
+bool SoftShutdownTracker::check() const {
+  Status status = getStatus();
+  if (!status.allClear) {
+    VPackBuilder builder;
+    toVelocyPack(builder, status);
+    // FIXME: Set to DEBUG level
+    LOG_TOPIC("ffeec", INFO, Logger::STARTUP)
+        << "Soft shutdown check said 'not all clear': "
+        << builder.slice().toJson() << ".";
+    return false;
+  }
+  LOG_TOPIC("ffeed", INFO, Logger::STARTUP)
+      << "Goal reached for soft shutdown, all ongoing tasks are terminated"
+         ", will now trigger the actual shutdown...";
+  initiateActualShutdown();
+  return true;
+}
+
+void SoftShutdownTracker::initiateActualShutdown() const {
+  Scheduler* scheduler = SchedulerFeature::SCHEDULER;
+  auto self = shared_from_this();
+  bool queued = scheduler->queue(RequestLane::CLUSTER_INTERNAL, [self] {
+    // Give the server 2 seconds to finish stuff
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+    self->_server.beginShutdown();
+  });
+  if (queued) {
+    return;
+  }
+  std::this_thread::sleep_for(std::chrono::seconds(2));
+  _server.beginShutdown();
+}
+
+void SoftShutdownTracker::toVelocyPack(VPackBuilder& builder,
+    SoftShutdownTracker::Status const& status) {
+  VPackObjectBuilder guard(&builder);
+  builder.add("softShutdownOngoing", VPackValue(status.softShutdownOngoing));
+  builder.add("AQLcursors", VPackValue(status.AQLcursors));
+  builder.add("transactions", VPackValue(status.transactions));
+  builder.add("pendingJobs", VPackValue(status.pendingJobs));
+  builder.add("doneJobs", VPackValue(status.doneJobs));
+  builder.add("pregelConductors", VPackValue(status.pregelConductors));
+  builder.add("lowPrioOngoingRequests", VPackValue(status.lowPrioOngoingRequests));
+  builder.add("lowPrioQueuedRequests", VPackValue(status.lowPrioQueuedRequests));
+  builder.add("allClear", VPackValue(status.allClear));
+}
+
+SoftShutdownTracker::Status SoftShutdownTracker::getStatus() const {
+  Status status;
+  status.softShutdownOngoing = _softShutdownOngoing.load(std::memory_order_relaxed);
```

Comment:
As mentoned below, I would make the `Status` object a bit more intelligent, so it is harder to use it wrong. For example, I would passing `softShutdownOngoing` as a mandatory ctor parameter, so it can't be forgotten. All other counters in the Status object I would initialize to 0 on object creation, and then only add to them.

Reply:
OK, will do.

Reply:
Done.